nextcloud/server
# Send images to Imaginary docker to generate previews

This is a preview provider for "image/*" mime types that streams the file data to an [Imaginary service](https://github.com/h2non/imaginary), which itself will generate the preview and stream it back.

### To test:

Using https://github.com/juliushaertl/nextcloud-docker-dev/pull/59

1. And enable the preview provider in config.php:
```
  'enabledPreviewProviders' => [
        'OC\Preview\MP3',
        'OC\Preview\TXT',
        'OC\Preview\MarkDown',
        'OC\Preview\OpenDocument',
        'OC\Preview\Krita',
        'OC\Preview\Imaginary',
  ],
```

2. Upload some jpeg pictures and see the previews rendered

### Some issues/topics:

#### Preview manager issues

- [x] <s>why is the maxX/maxY always 4096 even though the UI is requesting less than 400x400 ? (did I find a bug?)</s> Expected behavior, we crate a big preview first and then with GD resize it :/ Still something to improve and make it possible to also use imaginary to do the later resizing. Also worth investigating if having 4096x4096 has big previews make sense as most/many images have a smaller dimenssion.
- [x] BUG?  <s>even if setting only "Preview\Imaginary" in "enabledPreviewProviders", it still doesn't seem to be registered properly due to the other defaults. I was actually expecting "enabledPreviewProviders" to override the defaults.</s> Expected behavior the previews needs to be registered (in core with registerCorePreviews) and then enabled.

#### Imaginary preview topics

- [x] $response->getBody() is a string, but should be a stream => solved, post response was missing an arg
- [ ] optional perf improvement: if we know a file to be on local non-encrypted storage, could send the file path directly to the docker (when running on same instance)
- [x] gif not supported, it returns an error. might need more option
- [x] seems we do need multipart here, PUT not supported => fixed by sending POST with the actual mime type, not multipart

- [x] need to do some perf benchmarks and compare with the "native" way: https://github.com/nextcloud/server/pull/24166#issuecomment-728353823
- [x] do benchmarks with photos app (assuming it uses the PreviewManager)
